File: Pandacap.ActivityPub.Services/JsonLdExpansionService.fs
namespace Pandacap.ActivityPub.Services

open System
open Microsoft.Extensions.Caching.Memory
open Newtonsoft.Json.Linq
open JsonLD.Core
open Pandacap.ActivityPub.Services.Interfaces

type internal CustomDocumentLoader(cache: IMemoryCache) =
    inherit DocumentLoader()

    override _.LoadDocument (url: string): RemoteDocument = 
        let key = $"246d1dc3-a59e-49cc-a555-46d35635ac2d:{url}"

        match cache.TryGetValue(key) with
        | true, (:? RemoteDocument as cached) ->
            cached
        | _ ->
            cache.Set(
                key,
                base.LoadDocument(url),
                DateTimeOffset.UtcNow.AddHours(1))

type internal JsonLdExpansionService(cache: IMemoryCache) =
    interface IJsonLdExpansionService with
        member _.ExpandFirst(jObject: JObject) =
            let options = new JsonLdOptions(
                documentLoader = new CustomDocumentLoader(cache))

            try
                JsonLdProcessor.Expand(jObject, options) |> Seq.head
            with :? JsonLdError ->
                jObject["@context"] <- new JArray(
                    new JValue("https://www.w3.org/ns/activitystreams"),
                    new JValue("https://w3id.org/security/v1"))
                JsonLdProcessor.Expand(jObject, options) |> Seq.head

File: Pandacap.ActivityPub.JsonLd/JsonLdExpansionService.cs
using JsonLD.Core;
using Microsoft.Extensions.Caching.Memory;
using Newtonsoft.Json.Linq;
using Pandacap.ActivityPub.JsonLd.Interfaces;

namespace Pandacap.ActivityPub.JsonLd
{
    internal class JsonLdExpansionService(IMemoryCache memoryCache) : IJsonLdExpansionService
    {
        public JToken ExpandFirst(JObject jObject)
        {
            var options = new JsonLdOptions
            {
                documentLoader = new CustomDocumentLoader(memoryCache)
            };

            try
            {
                return JsonLdProcessor.Expand(jObject, options);
            }
            catch (JsonLdError)
            {
                jObject["@context"] = new JArray(
                    new JValue("https://www.w3.org/ns/activitystreams"),
                    new JValue("https://w3id.org/security/v1"));

                return JsonLdProcessor.Expand(jObject, options);
            }
        }

        private class CustomDocumentLoader(IMemoryCache memoryCache) : DocumentLoader
        {
            private const string CACHE_KEY_PREFIX = "246d1dc3-a59e-49cc-a555-46d35635ac2d";

            public override async Task<RemoteDocument> LoadDocumentAsync(string url)
            {
                var key = $"{CACHE_KEY_PREFIX}:{url}";

                if (memoryCache.TryGetValue(key, out var found) && found is RemoteDocument cached)
                    return cached;

                return memoryCache.Set(
                    key,
                    await base.LoadDocumentAsync(url),
                    DateTimeOffset.UtcNow.AddHours(1));
            }
        }
    }
}
